d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -302,19 +276,138 @@ private ClusterProperties toClusterProperties(Struct clusterProperties, long ver
         JsonFormat.printer().print(clusterProperties).getBytes(StandardCharsets.UTF_8), version);
   }
 
-  private UriProperties toUriProperties(String clusterName, Map<String, XdsD2.D2Node> uriDataMap)
+  private Map<String, UriProperties> toUriProperties(Map<String, XdsD2.D2Node> uriDataMap)
       throws InvalidProtocolBufferException, PropertySerializationException
   {
-    List<UriProperties> allUriProperties = new ArrayList<>();
+    Map<String, UriProperties> parsedMap = new HashMap<>();
 
-    for (XdsD2.D2Node d2Node : uriDataMap.values())
+    for (Map.Entry<String, XdsD2.D2Node> entry : uriDataMap.entrySet())
     {
+      XdsD2.D2Node d2Node = entry.getValue();
       UriProperties uriProperties = _uriPropertiesJsonSerializer.fromBytes(
           JsonFormat.printer().print(d2Node.getData()).getBytes(StandardCharsets.UTF_8), d2Node.getStat().getMzxid());
-      allUriProperties.add(uriProperties);
+      parsedMap.put(entry.getKey(), uriProperties);
+    }
+
+    return parsedMap;
+  }
+
+  private class UriPropertiesResourceWatcher implements XdsClient.D2NodeMapResourceWatcher
+  {
+    final String _clusterName;
+    final AtomicBoolean _init;
+    final long _initFetchStart;
+
+    Map<String, UriProperties> _currentData = new HashMap<>();
+
+    public UriPropertiesResourceWatcher(String clusterName)
+    {
+      _clusterName = clusterName;
+      _init = new AtomicBoolean();
+      _initFetchStart = System.nanoTime();
+    }
+
+    @Override
+    public void onChanged(XdsClient.D2NodeMapUpdate update)
+    {
+      if (_init.compareAndSet(false, true))
+      {
+        emitSDStatusInitialRequestEvent(_clusterName);
+      }
+
+      if (_uriEventBus != null)
+      {
+        try
+        {
+          Map<String, UriProperties> updates = toUriProperties(update.getNodeDataMap());
+          emitSDStatusUpdateReceiptEvents(updates);
+          _currentData = updates;
+          UriProperties mergedUriProperties = _uriPropertiesMerger.merge(_clusterName, _currentData.values());
+          _uriEventBus.publishInitialize(_clusterName, mergedUriProperties);
+          if (_dualReadStateManager != null)
+          {
+            _dualReadStateManager.reportData(_clusterName, mergedUriProperties, true);
+          }
+        }
+        catch (InvalidProtocolBufferException | PropertySerializationException e)
+        {
+          _log.error("Failed to parse D2 uri properties from xDS update. Cluster name: " + _clusterName, e);
+        }
+      }
+    }
+
+    @Override
+    public void onError(Status error)
+    {
+      notifyAvailabilityChanges(false);
```

Comment:
if the initial fetch failed, we need to emit the initialRequestEvent with failure, right?